src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -277,42 +302,66 @@ public void Tick(DateTimeOffset now)
                     Timeout();
                 }
             }
-            else
+        }
+
+        private void CheckForReadTimeout(long timestamp)
+        {
+            if (TimedOut || _timeoutTimestamp < long.MaxValue)
```

Comment:
I think this deserves a comment.

We don't check for read timeouts if there's a _timeoutTimestamp set? Why not? Isn't this taken care of by _readTimingEnabled?

If we really don't want to do this, It would probably be both more correct an easier to read as `if (TimedOut || Interlocked.Read(_timeoutTimestamp) != long.MaxValue)`